varname.py

```diff
@@ -19,18 +19,24 @@
 class VarnameRetrievingError(Exception):
     """When failed to retrieve the varname"""
 
-def varname(caller: int = 1, raise_exc: bool = True) -> Optional[str]:
+def varname(caller: int = 1,
+            multi_vars: bool = False,
+            raise_exc: bool = True) -> Optional[Union[str, Tuple[str]]]:
```

Comment:
```suggestion
            raise_exc: bool = True) -> Optional[Union[str, Tuple[Union[str, tuple]]]]:
```

What you really need is `TupleOrString = Union[str, Tuple["TupleOrString"]]` but that's not supported https://stackoverflow.com/questions/53845024/defining-a-recursive-type-hint-in-python

Reply:
Why wouldn't we just do: `Optional[Union[str, tuple]]`, since we can't expand `tuple` anyway.

Reply:
The only options are:

- `Optional[Union[str, tuple]]`
- `Optional[Union[str, Tuple[Union[str, tuple]]]]`
- `Optional[Union[str, Tuple[Union[str, Tuple[Union[str, tuple]]]]]]`
- ... and so on, replacing `tuple` with `Tuple[Union[str, tuple]]` each time

The first option (as you suggested) is fine. But some people might wonder "tuple of what?" and so the following layers add more information. It's hard to say if that's important.

I was saying that the ideal would be:

```python
TupleOrString = Union[str, Tuple["TupleOrString"]]
def varname(...) -> Optional[TupleOrString]
```

Since that raises an error in mypy it's not acceptable. It also doesn't actually offer any information in PyCharm so it's really not useful. But theoretically it's what should be supported one day and it's what we're trying to say.

The original `Optional[Union[str, Tuple[str]]]` is not correct because it suggests that if it's a tuple, it's a tuple of strings.

Your new suggestion is:

```python
TupleOrString = Union[str, Tuple[str]]
def varname(...) -> Optional[Union[TupleOrString, Tuple["TupleOrString"]]]
# inline the first TupleOrString
def varname(...) -> Optional[Union[Union[str, Tuple[str]], Tuple["TupleOrString"]]]
# flatten the nested union
def varname(...) -> Optional[Union[str, Tuple[str], Tuple["TupleOrString"]]]
# inline the second TupleOrString
def varname(...) -> Optional[Union[str, Tuple[str], Tuple[Union[str, Tuple[str]]]]]
# Tuple[str] is just a specific subtype of Tuple[Union[str, Tuple[str]]] so it's redundant
def varname(...) -> Optional[Union[str, Tuple[Union[str, Tuple[str]]]]]
# which is a more restrictive version of 
def varname(...) -> Optional[Union[str, Tuple[Union[str, tuple]]]]
```

In the end it's still incorrect because it hides that there can be more than two layers of tuples. Plus it's complicated and confusing. It's nothing like what I was saying because it's not recursive.

Reply:
See your point.
Then I will take the second one to balance off for now.